bib/cmd/bootc-image-builder/image.go

```diff
@@ -77,20 +75,74 @@ func Manifest(c *ManifestConfig) (*manifest.Manifest, error) {
 	}
 }
 
-func applyFilesystemCustomizations(customizations *blueprint.Customizations, c *ManifestConfig) error {
-	// Check the filesystem customizations against the policy and set
-	// if user configured
-	if err := blueprint.CheckMountpointsPolicy(customizations.GetFilesystems(), policies.OstreeMountpointPolicies); err != nil {
-		return err
-	}
-	if customFS := customizations.GetFilesystems(); len(customFS) != 0 {
-		allowedMounts := []string{"/", "/boot"}
-		for _, mp := range customFS {
-			if !slices.Contains(allowedMounts, mp.Mountpoint) {
-				return fmt.Errorf("cannot use custom mount points yet, found: %v", mp.Mountpoint)
-			}
+var (
+	// The mountpoint policy for bootc images is more restrictive than the
+	// ostree mountpoint policy defined in osbuild/images. It only allows /
+	// (for sizing the root partition) and custom mountpoints under /var but
+	// not /var itself.
+
+	// Since our policy library doesn't support denying a path while allowing
+	// its subpaths (only the opposite), we augment the standard policy check
+	// with a simple search through the custom mountpoints to deny /var
+	// specifically.
+	mountpointPolicy = pathpolicy.NewPathPolicies(map[string]pathpolicy.PathPolicy{
+		// allow all existing mountpoints to support size customizations
+		"/":     {Deny: false, Exact: true},
+		"/boot": {Deny: false, Exact: true},
+
+		// /var is not allowed, but we need to allow any subdirectories that
+		// are not denied below, so we allow it initially and then check it
+		// separately
+		"/var": {Deny: false},
+
+		// /var subdir denials
+		"/var/home":     {Deny: true},
+		"/var/lock":     {Deny: true}, // symlink to ../run/lock which is on tmpfs
+		"/var/mail":     {Deny: true}, // symlink to spool/mail
+		"/var/mnt":      {Deny: true},
+		"/var/roothome": {Deny: true},
+		"/var/run":      {Deny: true}, // symlink to ../run which is on tmpfs
+		"/var/srv":      {Deny: true},
+		"/var/usrlocal": {Deny: true},
+	})
+
+	mountpointMinimalPolicy = pathpolicy.NewPathPolicies(map[string]pathpolicy.PathPolicy{
+		// allow all existing mountpoints to support size customizations
+		"/":     {Deny: false, Exact: true},
+		"/boot": {Deny: false, Exact: true},
+	})
+)
+
+func checkMountpoints(filesystems []blueprint.FilesystemCustomization, policy *pathpolicy.PathTrie) error {
+	invalid := make([]string, 0)
+	for _, fs := range filesystems {
+		if err := policy.Check(fs.Mountpoint); err != nil {
+			invalid = append(invalid, fs.Mountpoint)
```

Comment:
It seems this check is discarding some information we get from the policy.Check() error that might be important to users and could lead to slightly misleading situations. Consider a customization that has `/boot/` instead of `/boot`. This will result in the following error (to the user): `The following custom mountpoints are not supported ["/boot/"]`. However the error from the underlying error from `Check()` is: `path must be canonical`.

Maybe instead (and tweak once https://github.com/osbuild/images/pull/853 is merged)?
```go
diff --git a/bib/cmd/bootc-image-builder/image.go b/bib/cmd/bootc-image-builder/image.go
index cc8be76..63a31f1 100644
--- a/bib/cmd/bootc-image-builder/image.go
+++ b/bib/cmd/bootc-image-builder/image.go
@@ -2,6 +2,7 @@ package main
 
 import (
        cryptorand "crypto/rand"
+       "errors"
        "fmt"
        "math"
        "math/big"
@@ -114,17 +115,17 @@ var (
 )
 
 func checkMountpoints(filesystems []blueprint.FilesystemCustomization, policy *pathpolicy.PathTrie) error {
-       invalid := make([]string, 0)
+       var errs []error
        for _, fs := range filesystems {
                if err := policy.Check(fs.Mountpoint); err != nil {
-                       invalid = append(invalid, fs.Mountpoint)
+                       errs = append(errs, err)
                }
                if fs.Mountpoint == "/var" {
-                       invalid = append(invalid, "/var")
+                       errs = append(errs, fmt.Errorf("path '/var' is not allowed"))
                }
        }
-       if len(invalid) > 0 {
-               return fmt.Errorf("The following custom mountpoints are not supported %+q", invalid)
+       if len(errs) > 0 {
+               return fmt.Errorf("The following errors occurred while setting up custom mountpoints:\n%w", errors.Join(errs...))
        }
        return nil
 }
diff --git a/bib/cmd/bootc-image-builder/image_test.go b/bib/cmd/bootc-image-builder/image_test.go
index 6dace68..6ff74d9 100644
--- a/bib/cmd/bootc-image-builder/image_test.go
+++ b/bib/cmd/bootc-image-builder/image_test.go
@@ -105,7 +105,7 @@ func TestCheckFilesystemCustomizationsValidates(t *testing.T) {
                                {Mountpoint: "/ostree"},
                        },
                        ptmode:      disk.RawPartitioningMode,
-                       expectedErr: `The following custom mountpoints are not supported ["/ostree"]`,
+                       expectedErr: "The following errors occurred while setting up custom mountpoints:\npath '/ostree ' is not allowed",
                },
                {
                        fsCust: []blueprint.FilesystemCustomization{
@@ -113,7 +113,7 @@ func TestCheckFilesystemCustomizationsValidates(t *testing.T) {
                                {Mountpoint: "/var"},
                        },
                        ptmode:      disk.RawPartitioningMode,
-                       expectedErr: `The following custom mountpoints are not supported ["/var"]`,
+                       expectedErr: "The following errors occurred while setting up custom mountpoints:\npath '/var' is not allowed",
                },
                {
                        fsCust: []blueprint.FilesystemCustomization{
@@ -121,7 +121,14 @@ func TestCheckFilesystemCustomizationsValidates(t *testing.T) {
                                {Mountpoint: "/var/data"},
                        },
                        ptmode:      disk.BtrfsPartitioningMode,
-                       expectedErr: `The following custom mountpoints are not supported ["/var/data"]`,
+                       expectedErr: "The following errors occurred while setting up custom mountpoints:\npath '/var/data ' is not allowed",
+               },
+               {
+                       fsCust: []blueprint.FilesystemCustomization{
+                               {Mountpoint: "/"},
+                               {Mountpoint: "/boot/"},
+                       },
+                       expectedErr: "The following errors occurred while setting up custom mountpoints:\npath must be canonical",
                },
        } {
                cust := &blueprint.Customizations{
```

```

```

Reply:
Thanks.  You're absolutely right of course.